File: problems/LeetCode/TopInterview150/canJump.ts
function canJump(nums: number[]): boolean {
  let pos = nums.length - 1;
  let jump = true;

  while (pos !== 0 && jump) {
    jump = false;
    for (let i = pos - 1; i >= 0; i--) {
      if (nums[i] >= pos - i) {
        pos = i;
        jump = true;
      }
    }
  }

  return pos === 0;
}

function canJump1(nums: number[]): boolean {
  let goal = nums.length - 1;

  for (let i = nums.length - 2; i >= 0; i--) {
    if (nums[i] + i >= goal) {
      goal = i;
    }
  }

  return goal === 0;
}

File: problems/LeetCode/TopInterview150/canJump.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:

        n = len(nums)
        goal = n - 1

        for i in range(n - 2, -1, -1):
            if i + nums[i] >= goal:
                goal = i

        return goal == 0



# Time: O(n)
# Space: O(1)